File: R/02.py
# See R code for puzzle info

import re

with open("../input/02-01.txt") as f:
  lines = f.readlines()

# 02-01 -------------------------------------------------------------------

def check1(line):
  
  match = re.search(
    "(?P<min>[^-]+)-(?P<max>[^ ]+) (?P<letter>[^:]+): (?P<password>.*)$",
    line
  )
  
  min = int(match.group("min"))
  max = int(match.group("max"))
  letter = match.group("letter")
  password = ''.join(chr for chr in match.group("password") if chr == letter)
  ct = len(password)
  
  return(
    (ct >= min) & (ct <= max)
  )

sum([ check1(line) for line in lines ])

# 02-02 -------------------------------------------------------------------

def check2(line):
  
  match = re.search(
    "(?P<pos1>[^-]+)-(?P<pos2>[^ ]+) (?P<letter>[^:]+): (?P<password>.*)$",
    line
  )
  
  pos1 = int(match.group("pos1"))
  pos2 = int(match.group("pos2"))
  letter = match.group("letter")
  password = match.group("password")
  
  return(
    ((password[pos1-1] == letter) |
     (password[pos2-1] == letter)) &
     (password[pos1-1] != password[pos2-1])
  )

sum([ check2(line) for line in lines ])

File: js/02.js
// See R code for the problem texts

// 02-01

var fs = require("fs")

input = fs.readFileSync("../input/02-01.txt", "utf-8")
  .split("\n")


const re1 = /(?<min>[^-]+)-(?<max>[^ ]+) (?<letter>[^:]+): (?<password>.*)$/;

res = input.map((line) => {

  g = re1.exec(line).groups

  g.min = parseInt(g.min)
  g.max = parseInt(g.max)

  ltr_re = new RegExp("[^" + g.letter + "]", "g")
  g.password = g.password.replace(ltr_re, "")

  return(
    (g.password.length >= g.min) &
     (g.password.length <= g.max)
  )

}).reduce((a, b) => a + b)

console.log(res)

// 02-02

const re2 = /(?<pos1>[^-]+)-(?<pos2>[^ ]+) (?<letter>[^:]+): (?<password>.*)$/;

res = input.map((line) => {

  g = re2.exec(line).groups

  g.pos1 = parseInt(g.pos1)
  g.pos2 = parseInt(g.pos2)

  return(((g.password[g.pos1-1] == g.letter) |
   (g.password[g.pos2-1] == g.letter)) &
   (g.password[g.pos1-1] != g.password[g.pos2-1]))

}).reduce((a, b) => a + b)

console.log(res)
